homeassistant/components/mqtt/__init__.py

```diff
@@ -369,7 +369,9 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     mqtt_data.client = MQTT(hass, entry, conf)
     # Restore saved subscriptions
     if mqtt_data.subscriptions_to_restore:
-        mqtt_data.client.subscriptions = mqtt_data.subscriptions_to_restore
+        mqtt_data.client.async_restore_tracked_subscriptions(
+            mqtt_data.subscriptions_to_restore
+        )
```

Comment:
I seems we can remove this. Also we do not need to store subscriptions when unloading the entry to be able to restore them. This part of code is stale, as all subscriptions are renewed when the entry starts.

Reply:
May we need some extra tests to prove that subscriptions are cleaned up though, I believe we have some of those when unloading the entry.

Reply:
Subscriptions done externally from other integrations (or via the debug panel on the MQTT config entry page), are left over after unloading the entry. I have tested this in a real environment. Also added a test.